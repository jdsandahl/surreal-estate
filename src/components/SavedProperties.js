import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import FavouriteCard from "./FavouriteCard";
import Alert from "./Alert";
import deleteFavourite from "../services/deleteFavourite";
import "../styles/SavedProperties.css";

const SavedProperties = ({ userId }) => {
  const initialState = {
    favourites: [],
    loading: false,
    alert: {
      message: "",
      isSuccess: true,
    },
  };

  const [favourites, setFavourites] = useState(initialState.favourites);
  const [loading, setLoading] = useState(initialState.loading);
  const [alert, setAlert] = useState(initialState.alert);

  useEffect(() => {
    setLoading(true);

    axios
      .get(
        `http://localhost:4000/api/v1/Favourite?query={"fbUserId":"${userId}"}&populate=propertyListing`
      )
      .then(({ data }) => {
        setFavourites(data);
      })
      .catch((err) => {
        setAlert({
          message: "Server Error: Properties not found, please try again later",
          isSuccess: false,
        });
      });
    setLoading(false);
  }, [userId, favourites]);

  const clearAlert = () => {
    setAlert(initialState.alert);
  };

  const handleRemoveFavourite = async (favouriteId) => {
    console.log(favouriteId);
    setAlert(await deleteFavourite(favouriteId));
    setTimeout(clearAlert, 3000);
  };

  return (
    <div className="saved-properties" data-testid="saved-properties">
      {loading && alert.isSuccess === true && (
        <div className="favourites__loading">loading</div>
      )}
      <Alert message={alert.message} success={alert.isSuccess} />
      <div className="favourites-list">
        {favourites.map((favourite) => (
          <FavouriteCard
            key={favourite._id}
            {...favourite}
            onRemoveFavourite={handleRemoveFavourite}
          />
        ))}
      </div>
    </div>
  );
};

SavedProperties.propTypes = {
  userId: PropTypes.string.isRequired,
};

export default SavedProperties;
