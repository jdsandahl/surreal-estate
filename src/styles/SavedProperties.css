.saved-properties {
    display: flex;
    flex-direction: column;
    background-color: var(--secondary-light-color);
    height: 100vh;
}