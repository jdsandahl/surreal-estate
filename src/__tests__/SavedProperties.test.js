import React from 'react';
import { render } from '@testing-library/react';
import SavedProperties from '../components/SavedProperties';

describe('SavedProperties', () => {
    test('it renders correctly against a snapshot', () => {
        const mockUserId = "mockId"
        const { asFragment } = render (<SavedProperties userId={mockUserId}/>)
        
        expect(asFragment()).toMatchSnapshot();
    });

    test('component elements render correctly', () => {
        const mockUserId = "mockId"
        const { getByTestId } = render (<SavedProperties userId={mockUserId}/>)
        
        const component = getByTestId("saved-properties");

        expect(component).toBeInTheDocument();
    });
});