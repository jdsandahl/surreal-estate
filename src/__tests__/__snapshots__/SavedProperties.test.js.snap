// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedProperties it renders correctly against a snapshot 1`] = `
<DocumentFragment>
  <div
    class="saved-properties"
    data-testid="saved-properties"
  >
    <div
      class="loading"
    >
      <div
        class="indeterminate"
      />
    </div>
    <div
      class="favourites-list"
    />
  </div>
</DocumentFragment>
`;
